orbit/cmd/orbit/orbit.go

```diff
@@ -481,64 +499,111 @@ func main() {
 }
 
+// execute makes sure the fleet-desktop application is running.
+//
+// We have to support the scenario where the user closes its sessions (log out).
+// To support this, we add retries to execuser.Run. Basically retry execuser.Run until it succeds,
+// which will happen when the user logs in.
+// Once fleet-desktop is started, the process is monitored (user processes get killed when the user
+// closes all its sessions).
+//
+// NOTE(lucas): This logic could be improved to detect if there's a valid session or not first.
 func (d *desktopRunner) execute() error {
-	log.Info().Str("path", d.appPath).Msg("opening")
+	defer close(d.executeDoneCh)
+
+	log.Info().Str("path", d.desktopPath).Msg("opening")
 	url, err := url.Parse(d.fleetURL)
 	if err != nil {
 		return fmt.Errorf("invalid fleet-url: %w", err)
 	}
 	url.Path = path.Join(url.Path, "device", d.deviceAuthToken)
-	opts := []open.AppOption{
-		open.AppWithEnv("FLEET_DESKTOP_DEVICE_URL", url.String()),
-		open.AppWithEnv("FLEET_DESKTOP_DEVICE_API_TEST_PATH", path.Join("api", "latest", "fleet", "device", d.deviceAuthToken)),
+	opts := []execuser.Option{
+		execuser.WithEnv("FLEET_DESKTOP_DEVICE_URL", url.String()),
+		execuser.WithEnv("FLEET_DESKTOP_DEVICE_API_TEST_PATH", path.Join("api", "latest", "fleet", "device", d.deviceAuthToken)),
 	}
 	if d.insecure {
-		opts = append(opts, open.AppWithEnv("FLEET_DESKTOP_INSECURE", "1"))
+		opts = append(opts, execuser.WithEnv("FLEET_DESKTOP_INSECURE", "1"))
 	}
-	if err := open.App(d.appPath, opts...); err != nil {
-		return fmt.Errorf("open desktop app: %w", err)
+
+	for {
+		// First retry logic to start fleet-desktop.
+		if done := retry(30*time.Second, d.interruptCh, func() bool {
+			// Orbit runs as root user on Unix and as SYSTEM (Windows Service) user on Windows.
+			// To be able to run the desktop application (mostly to register the icon in the system tray)
+			// we need to run the application as the login user.
+			// Package execuser provides multi-platform support for this.
+			if err := execuser.Run(d.desktopPath, opts...); err != nil {
+				log.Debug().Err(err).Msg("execuser.Run")
+				return true
+			}
+			return false
+		}); done {
+			return nil
+		}
+
+		// Second retry logic to monitor fleet-desktop.
+		if done := retry(30*time.Second, d.interruptCh, func() bool {
+			switch _, err := getProcessByName(constant.DesktopAppExecName); {
+			case err == nil:
+				return true // all good, process is running, retry.
+			case errors.Is(err, errProcessNotFound):
+				log.Debug().Msgf("%s process not running", constant.DesktopAppExecName)
+				return false // process is not running, do not retry.
+			default:
+				log.Debug().Err(err).Msg("getProcessByName")
+				return true // failed to get process by name, retry.
+			}
+		}); done {
+			return nil
+		}
 	}
+}
 
-	// Monitor the fleet-desktop application.
-	ticker := time.NewTicker(10 * time.Second)
+func retry(d time.Duration, done chan struct{}, fn func() bool) bool {
```

Comment:
I wonder if we should use https://pkg.go.dev/github.com/cenkalti/backoff#Retry or related functions since we already have that dependency and use it in some other places?

Reply:
+1, will consider using it when working on Linux.